src/rust/en.mangairo/src/parser.rs

```diff
@@ -0,0 +1,354 @@
+		13 => "15", // "Historical",
+		14 => "16", // "Horror",
+		15 => "45", // "Isekai",
+		16 => "17", // "Josei",
+		17 => "44", // "Manhua",
+		18 => "43", // "Manhwa",
+		19 => "19", // "Martial arts",
+		20 => "20", // "Mature",
+		21 => "21", // "Mecha",
+		22 => "22", // "Medical",
+		23 => "24", // "Mystery",
+		24 => "25", // "One shot",
+		25 => "47", // "Pornographic",
+		26 => "26", // "Phychological",
+		27 => "27", // "Romance",
+		28 => "28", // "School life",
+		29 => "29", // "Sci fi",
+		30 => "30", // "Seinen",
+		31 => "31", // "Shoujo",
+		32 => "32", // "Shoujo ai",
+		33 => "33", // "Shounen",
+		34 => "34", // "Shounen ai",
+		35 => "35", // "Slice of Life",
+		36 => "36", // "Smut",
+		37 => "37", // "Sports",
+		38 => "38", // "Supernatural",
+		39 => "39", // "Tragedy",
+		40 => "40", // "Webtoons",
+		41 => "41", // "Yaoi",
+		42 => "42", // "Yuri"
+		_ => "all",
+	};
+
+	// State
+	let status = match status_filter {
+		0 => "all",
+		1 => "ongoing",
+		2 => "completed",
+		_ => "all",
+	};
+
+	let page = &page.to_string();
+	format!("{BASE_URL}/manga-list/type-{sort}/ctg-{ctg}/state-{status}/page-{page}")
+}
+
+pub fn parse_incoming_url_manga_id(url: String) -> Option<String> {
+	// https://chap.mangairo.com/story-pn279847
+	// https://chap.mangairo.com/story-pn279847/chapter-52
+	let mut parts: Vec<&str> = url.split('/').collect();
+	if parts.len() >= 4 {
+		parts.truncate(4);
+	}
+
+	Some(parts.join("/"))
+}
+
+pub fn parse_incoming_url_chapter_id(url: String) -> Option<String> {
+	// https://chap.mangairo.com/story-pn279847/chapter-52
+	let parts: Vec<&str> = url.split('/').collect();
+	if parts.len() >= 4 {
+		let chapter_id = parts[4];
+		return Some(format!("{}", chapter_id));
+	}
+
+	None
+}
+
+// HELPER FUNCTIONS
+
+pub fn urlencode(string: String) -> String {
+	let mut str = string.to_lowercase();
+
+	let match_a = [
+		'à', 'á', 'ạ', 'ả', 'ã', 'â', 'ầ', 'ấ', 'ậ', 'ẩ', 'ẫ', 'ă', 'ằ', 'ắ', 'ặ', 'ẳ', 'ẵ',
+	];
+	let match_e = ['è', 'é', 'ẹ', 'ẻ', 'ẽ', 'ê', 'ề', 'ế', 'ệ', 'ể', 'ễ'];
+	let match_i = ['ì', 'í', 'ị', 'ỉ', 'ĩ'];
+	let match_o = [
+		'ò', 'ó', 'ọ', 'ỏ', 'õ', 'ô', 'ồ', 'ố', 'ộ', 'ổ', 'ỗ', 'ơ', 'ờ', 'ớ', 'ợ', 'ở', 'ỡ',
+	];
+	let match_u = ['ù', 'ú', 'ụ', 'ủ', 'ũ', 'ư', 'ừ', 'ứ', 'ự', 'ử', 'ữ'];
+	let match_y = ['ỳ', 'ý', 'ỵ', 'ỷ', 'ỹ'];
+	let match_d = "đ";
+	let match_symbols = [
+		'!', '@', '%', '^', '*', '(', ')', '+', '=', '<', '>', '?', '/', ',', '.', ':', ';', '\'',
+		' ', '"', '&', '#', '[', ']', '~', '-', '$', '|', '_',
+	];
+
+	str = str.replace(match_a, "a");
+	str = str.replace(match_e, "e");
+	str = str.replace(match_i, "i");
+	str = str.replace(match_o, "o");
+	str = str.replace(match_u, "u");
+	str = str.replace(match_y, "y");
+	str = str.replace(match_d, "d");
+	str = str.replace(match_symbols, "_");
+	str = replace_consecutive_underscores(str);
+	str = str.trim_matches('_').to_string();
+
+	str
+}
+
+fn replace_consecutive_underscores(input: String) -> String {
```

Comment:
I'm a little confused. Isn't this the same as `.replace("__", "_")`?

Reply:
That would collapse three underscores `___` to `_`? I was thinking it would not, but now that I think about it again it makes sense it would.